File: pyutil/_merkle_tree.py
from json import dumps
from ..poseidon.poseidon import poseidon

class MerkleTree:
    
    def __init__(self, depth):
        self.tree = [dict() for _ in range(depth)]
        self.depth = depth
        self._size = 0
    
    def __len__(self):
        return self._size
    
    def root(self):
        assert self._size > 0, "tree is empty"
        return self.tree[0][0]
    
    def addElement(self, x):
        self.tree[self.depth-1][self._size] = x
        index = self._size
        for d in reversed(range(1, self.depth)):
            if (index % 2 == 0):
                self.tree[d][index+1] = self.tree[d][index]
            index //= 2
            self.tree[d-1][index] = poseidon(
                self.tree[d][2*index], 
                self.tree[d][2*index+1]
            )
        self._size += 1
    
    def proof(self, i):
        assert 0 <= i < self._size
        merkle_proof = [None] * (self.depth - 1)
        for j in reversed(range(1, self.depth)):
            d = 1 if i % 2 == 0 else -1
            l, r = self.tree[j][i], self.tree[j][i+d]
            merkle_proof[j-1] = (l, r) if d == 1 else (r, l)
            i //= 2
        return merkle_proof
    
    @staticmethod
    def verifyProof(leaf, merkle_proof, merkle_root):
        assert leaf in merkle_proof[-1]
        for i in reversed(range(len(merkle_proof))):
            hsh = poseidon(*merkle_proof[i])
            if i > 0:
                assert (hsh in merkle_proof[i-1]), (
                    f"hashing at level {i} doesn't produce"
                    f"a hash at level {i-1}")
            else:
                assert (hsh == merkle_root), (
                    "end hash doesn't match the merkle root")

File: tsutil/merkleTree.ts
import { assert } from "chai"
import { postreidon } from "./poseidon/poseidon"

export class MerkleTree {
    tree: any[] = []
    depth: number
    _size: number

    constructor(depth: number) {
        for (let i = 0; i < depth; i++) 
            this.tree.push({})
        this.depth = depth
        this._size = 0
    }

    root(): string {
        return this.tree[0][0]
    }

    addElement(x: string) {
        this.tree[this.depth-1][this._size] = x
        let index = this._size
        for (let d = this.depth-1; d > 0; d--) {
            if (index % 2 == 0) 
                this.tree[d][index+1] = this.tree[d][index]
            index >>= 1
            this.tree[d-1][index] = postreidon([
                this.tree[d][2*index],
                this.tree[d][2*index+1]
            ])
        }
        this._size += 1
    }

    proof(i: number): string[][] {
        let merkleProof: string[][] = []
        for (let i = 0; i < this.depth-1; i++) 
            merkleProof.push([""])
        for (let j = this.depth-1; j > 0; j--) {
            const d = (i % 2 == 0) ? 1 : -1
            const l = this.tree[j][i]
            const r = this.tree[j][i+d]
            merkleProof[j-1] = (d == 1) ? [l, r] : [r, l]
            i >>= 1
        } 
        return merkleProof
    }

    verifyProof(
        leaf: string, merkleProof: string[][], 
        merkleRoot: string
    ) {
        const l = merkleProof.length
        assert(
            leaf == merkleProof[l-1][0] || 
            leaf == merkleProof[l-1][1])
        for (let i = l-1; i >= 0; i--) {
            const hsh = postreidon(merkleProof[i])
            if (i > 0) assert(
                hsh == merkleProof[i-1][0] ||
                hsh == merkleProof[i-1][1])
            else assert(hsh == merkleRoot)
        }
    }
}